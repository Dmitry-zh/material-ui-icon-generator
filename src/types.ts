import { HTMLAttributes } from 'react'
import { HTMLElement, Node } from 'node-html-parser'

export type RawTagName = string | null

export type WithRawTagName<T> = T & { rawTagName: RawTagName }

export type ParsedHtml = WithRawTagName<HTMLElement> | Node

export type SvgTree = {
  tag: RawTagName
  attributes: HTMLAttributes<SVGElement> | null
  children: SvgTree[]
}

export type GenCallback = () => Promise<string | undefined>
