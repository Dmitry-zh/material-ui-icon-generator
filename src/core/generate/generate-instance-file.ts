import path from 'path'
import fs from 'fs'

import { path as rootPath } from 'app-root-path'

export const generateInstanceFile = async () => {
  let fileContent = ''
  const outputDir = path.join(rootPath, 'output')

  await new Promise((resolve, reject) => {
    fs.readdir(outputDir, async (err, files) => {
      if (err) reject(err.message)

      files.forEach((file, index) => {
        const fileName = path.parse(file).name
        fileContent += `export {default as ${fileName}} from './${fileName}'`
        fileContent += '\n'
        if (index === files.length - 1) {
          resolve('done')
        }
      })
    })
  })

  return fs.writeFileSync(
    path.join(outputDir, 'index.js'),
    Buffer.from(fileContent),
  )
}
