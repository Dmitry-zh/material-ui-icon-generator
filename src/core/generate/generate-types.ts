import path from 'path'
import fs from 'fs'

import { OUT_DIR } from '~/constants'

export const generateTypes = async () => {
  let fileContent = ''

  await new Promise((resolve, reject) => {
    fs.readdir(OUT_DIR, async (err, files) => {
      if (err) reject(err.message)

      files.forEach((file, index) => {
        const fileName = path.parse(file).name
        fileContent += `export {default as ${fileName}} from '@mui/material/SvgIcon'`
        fileContent += '\n'
        if (index === files.length - 1) {
          resolve('done')
        }
      })
    })
  })

  return fs.writeFileSync(
    path.join(OUT_DIR, 'index.d.ts'),
    Buffer.from(fileContent),
  )
}
