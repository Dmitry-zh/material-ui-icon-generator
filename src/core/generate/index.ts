export { generateIconFile } from './generate-icon-file'
export { generateInstanceFile } from './generate-instance-file'
export { generateTypes } from './generate-types'
export { processOutputDir } from './process-output-dir'
