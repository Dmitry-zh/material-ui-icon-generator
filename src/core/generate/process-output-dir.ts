import fs from 'fs'

import { OUT_DIR } from '~/constants'
import { logger } from '~/util'

const createOutputDir = () => {
  fs.mkdirSync(OUT_DIR)

  return logger.info('OUTPUT DIR WAS CREATED')
}

export const processOutputDir = () => {
  return new Promise((resolve, reject) => {
    if (!fs.existsSync(OUT_DIR)) {
      return resolve(createOutputDir())
    }

    fs.rm(OUT_DIR, { force: true, recursive: true }, (error) => {
      if (error) {
        return reject(
          logger.error('ERROR OCCURRED CREATING OUT DIR: %s', error.message),
        )
      }

      return resolve(createOutputDir())
    })
  })
}
