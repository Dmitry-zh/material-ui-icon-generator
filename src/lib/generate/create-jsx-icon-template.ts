import { SvgTree } from '../../types'

interface CreateJsxIconTemplateParams {
  iconName: string
  elements: SvgTree['children']
}

export const createJsxIconTemplate = ({
  iconName,
  elements,
}: CreateJsxIconTemplateParams): string => {
  return `
import {jsx} from 'react/jsx-runtime'
import {createSvgIcon} from '@mui/material'

export default createSvgIcon([
${elements.map(
  (el, index) => `
jsx(
 '${el.tag}',
 ${JSON.stringify(el.attributes)},
 '${index}'
)
`,
)}
], '${iconName}')
`
}
