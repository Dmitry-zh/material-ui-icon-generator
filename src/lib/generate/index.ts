export { generateIconFile } from './generate-icon-file'
export { generateInstanceFile } from './generate-instance-file'
export { processOutputDir } from './process-output-dir'
