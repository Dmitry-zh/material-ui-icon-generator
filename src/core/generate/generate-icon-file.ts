import path from 'path'
import fs from 'fs'

import { SvgTree } from '~/types'
import { logger } from '~/util'
import { OUT_DIR } from '~/constants'

import { pascalCase } from '../util'

import { createJsxIconTemplate } from './create-jsx-icon-template'

export const generateIconFile = async (
  tree: SvgTree['children'],
  iconFileName: string,
) => {
  return new Promise((resolve, reject) => {
    const name = pascalCase(path.parse(iconFileName).name)

    const outputDir = path.join(OUT_DIR, name + '.js')
    const template = createJsxIconTemplate({ iconName: name, elements: tree })

    fs.writeFile(outputDir, Buffer.from(template), (error) => {
      if (error) {
        return reject(
          logger.error(
            'ERROR OCCURRED WRITING ICON: %s on file %s',
            error.message,
            name,
          ),
        )
      }
      return resolve(name)
    })
  })
}
