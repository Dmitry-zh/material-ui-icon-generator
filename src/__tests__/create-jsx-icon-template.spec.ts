import { describe, expect, it } from 'vitest'

import { createJsxIconTemplate } from '~/core/generate/create-jsx-icon-template'

import { SvgTree } from '../types'

import { sunTree } from './fixtures/sun-tree'

describe('generate lib tests', () => {
  it('should generate correct template', async () => {
    expect(
      createJsxIconTemplate({
        elements: sunTree as SvgTree['children'],
        iconName: 'sun',
      }),
    ).toMatchSnapshot()
  })
})
