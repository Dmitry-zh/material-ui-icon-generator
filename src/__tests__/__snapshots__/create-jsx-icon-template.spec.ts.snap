// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`generate lib tests > should generate correct template 1`] = `
"
import {jsx} from 'react/jsx-runtime'
import createSvgIcon from '@mui/material/utils/createSvgIcon'

export default createSvgIcon([

jsx(
 'circle',
 {\\"cx\\":\\"12\\",\\"cy\\":\\"12\\",\\"r\\":\\"5\\"},
 '0'
)
,
jsx(
 'path',
 {\\"d\\":\\"M12 1v2M12 21v2M4.2 4.2l1.4 1.4M18.4 18.4l1.4 1.4M1 12h2M21 12h2M4.2 19.8l1.4-1.4M18.4 5.6l1.4-1.4\\"},
 '1'
)

], 'sun')
"
`;
